google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/BitReversedSequenceStyleGenerator.java

```diff
@@ -0,0 +1,169 @@
+      this.qualifiedName = qualifiedTableName;
+      this.valueColumnNameIdentifier = valueColumnNameIdentifier;
+      this.initialValue = initialValue;
+    }
+
+    @Override
+    public void registerExportables(Database database) {
+      super.registerExportables(database);
+      // Replace the init command for the table-backed sequence.
+      // Hibernate by default generates an 'insert into table_name values (?)' statement.
+      // That is not supported by Cloud Spanner, as Cloud Spanner requires the insert statement to
+      // include the column name(s) that are being used in the insert statement.
+      final Namespace namespace =
+          database.locateNamespace(qualifiedName.getCatalogName(), qualifiedName.getSchemaName());
+      Table table = namespace.locateTable(qualifiedName.getObjectName());
+      if (table != null) {
+        Dialect dialect = jdbcEnvironment.getDialect();
+        String valueColumnNameText = valueColumnNameIdentifier.render(dialect);
+        table.addInitCommand(
+            context ->
+                new ReplaceInitCommand(
+                    "insert into "
+                        + context.format(table.getQualifiedTableName())
+                        + " ("
+                        + valueColumnNameText
+                        + ") values ( "
+                        + initialValue
+                        + " )"));
+      }
+    }
+  }
+
+  /**
+   * Acts as a replacement for other {@link InitCommand}.
+   *
+   * <ol>
+   *   <li>If the list contains at least one {@link ReplaceInitCommand} and at least one normal
+   *       {@link InitCommand}, then all normal {@link InitCommand}s will be ignored during
+   *       execution and only the {@link ReplaceInitCommand}s will be executed.
+   *   <li>If the list only contains {@link ReplaceInitCommand}s, nothing will be executed.
+   *   <li>If the list only contains normal {@link InitCommand}s, all normal {@link InitCommand}s
+   *       will be executed as normal.
+   * </ol>
+   */
+  public static class ReplaceInitCommand extends InitCommand {
+    public ReplaceInitCommand(String... initCommands) {
+      super(initCommands);
+    }
+  }
+
+  @Override
+  protected DatabaseStructure buildTableStructure(
+      Type type,
+      Properties params,
+      JdbcEnvironment jdbcEnvironment,
+      QualifiedName sequenceName,
+      int initialValue,
+      int incrementSize) {
+    final Identifier valueColumnName = determineValueColumnName(params, jdbcEnvironment);
+    return new SpannerSequenceTableStructure(
+        jdbcEnvironment,
+        sequenceName,
+        valueColumnName,
+        initialValue,
+        incrementSize,
+        type.getReturnedClass());
+  }
+
+  /**
+   * Generates a new ID. This uses the normal sequence strategy, but the returned ID is bit-reversed
+   * before it is returned to the application.
+   */
+  @Override
+  public Serializable generate(SharedSessionContractImplementor session, Object object)
```

Comment:
Can you add a test to check the sequence is actually reversed?

Reply:
Done